modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java

```diff
@@ -59,6 +83,28 @@ public CompletableFuture<Object> processRequest(ReplicaRequest request) {
                 replicaGrpId);
 
         return listener.invoke(request);
+    }
+
+    /**
+     * Whether this replica on this node is primary.
+     *
+     * @return Whether is primary.
+     */
+    public boolean isPrimary() {
+        // TODO https://issues.apache.org/jira/browse/IGNITE-17256
+        return raftClient.refreshAndGetLeaderWithTerm()
+            .thenApply(leaderAndTerm -> clusterNodeResolver.apply(leaderAndTerm.get1().address()).equals(localNodeSupplier.get()))
+            .join();
+    }
+
+    /**
+     * Propagates safe time on non-primary replicas, if this replica is primary.
+     */
+    public void propagateSafeTime() {
+        if (!isPrimary()) {
+            return;
+        }
 
+        raftClient.run(new SafeTimeSyncCommand());
```

Comment:
I think, that the Replica can only to invoke a replica listener. It shouldn't invoke RAFT as is.

Reply:
As we discussed, it will be better to invoke replica listener here (with specific SafeImeSycnReplicaRequest) and the listener invokes RAFT. Because the approach doesn't add a particular protocol reference to the replication layer.